### DISCUSSION TITLE
Weekly AI update

### DISCUSSION THREAD

**Discussion by lemonteaa** [hkitsmallpotato/let-ai-dev#13] [General]
At the pace this is going, weekly updates seems about the right frequency.

(no promise though - I'm occupied with other tasks and likely don't have the bandwidth to do this with any consistency)

Note: Focus on LLM + open source, though related developments in the external enviornment may also be mentioned if relevant.

**Comment by lemonteaa** [comment]
# Summary as of 26th May 2023
- qlora - it is now possible to fine-tune LLM directly on quantized model ( => reduced memory requirement for fine-tune).
  - This results in the Guanaco model, which claims to have high output quality
- `llama.cpp` supports AMD gpu now
- OpenLlama, an effort to reproduce llama foundation model but with a permissive license, has released model checkpoint (i.e. intermediate result) at 600B tokens
- `bitsandbytes` will soon release supports for 4 bit quantization, providing a HF native method (which would open this technique to more models in the HF ecosystem?)
- The code base for the lottery ticket hypothesis is also planned to be released by Meta.
  - The **lottery ticket hypothesis** refers to a paper saying that they found most of the actual functionality/intelligence of a LLM is concentrated in a small portion of neurons (it's like doing a lucky draw repeatedly with large number of neural units and one of them luck out into an intelligent configuration). In particular if we can identify the useful part, this may enable shrinking models by 80 or even 90+ % without substantial quality loss

**Comment by lemonteaa** [comment]
# Summary as of 16th Jun 2023

- **Landmark** is being implemented in some open source models that unlock longer context length
  - Perhaps in response to this/feeling the pressure, OpenAI announced the availability of long context length model (among other things) at almost the same time.
- `WizardCoder` is released and claims to have vastly improved on Coding tasks compared to all other existing OSS models, even beating Claude+ (a commercial/propietrary model)
  - The context is that there was an evaluation done on coding tasks specifically on Open source Models against GPT3.5/4 and it is found that all models have a huge gap/lags behind a lots. WizardCoder may pave the way to close in on this gap.
- **SqeezeLLM** is a new quantization method that promises lossless compression at up to 3bits and improved quality at same memory constraints
  - Similar proposal in terms of things promised include the **SPQR quantisation** scheme previously
- `llama.cpp` have implemented full GPU acceleration and now it can beat GPTQ sometimes when you have enough VRAM to fit the whole model in GPU (previously the recommendation is llama.cpp/ggml if you can't fit the whole model, but gptq if you can)
llama.cpp also seems to have started adding code to enable training LLMs.
- `operand/agency` is a library for application development on LLM, with a focus on the top/metalayer on top of langchain - it is best suited for having a community of LLM agents + tools + human communicating and cooperating with each others
- The EU have passed the AI act, which regulates unacceptable and high risk applications of AI. For LLM, the act generally requires transparency in the training dataset used, clearly indicating to end user when and where LLM is used, and to have safety measures when developing end user facing products
  - It will then enter the trialogue phase, and it's expected the law would take effect from the end of this year to a few year down the line

**Comment by lemonteaa** [comment]
# Other developments up to 16th June

(Some of these are older news)

- StarChat beta is now out, it took a different path: StarCodeBase (the one that only has python) => finetune with 80+ programming language dataset to get StarCode Plus => instruct finetune with guanaco, manticore etc instead of dolly + openassistent
- llama.cpp is funded and ggml become a startup company, also tinygrad is funded with 5M seed round
- Someone (seems to be one of the ex-github CTO etc) privately owns a large H100 cluster and is offering it to select startups

----

(Notice that I wrote all these in a rush and some factual details may be off, sorry in advance about that)

**Comment by lemonteaa** [comment]
# Summary as of 23rd June 2023
This week being a bit busy, I may have missed something.

- Vicuna 1.3 as well as SqueezeLLM (new quantisation method) is released
- Microsoft claims that it is possible to beat all the open source model on coding tasks (including `WizardCoder` which is already quite capable?), falling just behind GPT4, through high quality dataset. You don't even need big model - something as small as < 3B can do it (taking inspiration from the `tinystory` model). They call it `phi-1` and say they intend to release it soon. (They made the same claim on `Orca`, but haven't fulfilled their promise to release yet)
- There are rumour leak of GPT4 architecture - according to someone involved in pytorch, GPT4 is actually 8 x 220B models using something like a "Mixture of Experts" and their output are then selected before being presented to user. If true, this may imply that OSS is not too hopeless in catching up as the model scaling may not be as critical
- Finally, someone claims to did a hack on the inference code of `llama.cpp` and are then able to increase the context length to 8k without too much quality/intelligence degradation - and this is *without* doing any kind of training whatsoever on the model (!). The idea is that the positional embedding is a mistake that's too late to fix, and due to over-fitting the model has an ingrained idea to only accept inputs sequence number within the range. So the hack is "interpolate instead of extrapolate" - we make the token number as "1, 1.25, 1.5, 1.75, 2" instead of "1, 2, 3, 4" (!!)
- MPT-30B foundation model is released:
  - Apache 2.0 license
  - 8k context length
  - trained on 1T token
  - coding focus (seems to be)
- Vicuna-**33B** v1.3 is also released

**Comment by lemonteaa** [comment]
# Summary as of 6th July 2023

Sorry for the delay. Below is a quick draft, so the details and facts may be completely wrong. Keeping in mind this limitation, here we are:

## Non LLM news

- modelscope XL, an AI that can generate video based on text prompt, is released
- MotionGPT is a multimodal model that can understand 3D motion data and interface with natural language text

## LLM news

**New models and dataset**

- Saleforce released X-Gen
- OpenOrca is a dataset that may help to reproduces Microsoft's Orca paper as Microsoft seems to not be releasing them (despite promises to?)
  - Note that this is different from orca-mini, which is a fine-tuned model, again with the aim to reproduces the Microsoft Orca paper's result
  - However, this release seems to be mired in controversy and drama - such as claim of stealing credits and other issues
- orca-mini V2 is also released

**Advances and Notes**

- The new Nvidia driver aggressively offload VRAM in GPU into main memory, which causes regression in the performance of GPU only LLM inference softwares. Nvidia seems to be aware of the issue and is working on a fix, but in the mean time the advice is to downgrade or hold at a version before the regression
- The biggest news recently is that the embedding interpolation hack to extend context length limit without doing anything to the model itself (2k to 8k), become really big and is the majority of open source's focus this and last week.
  - We see this idea ethusiastically adopted and deployed rapidly and widely. So going forward if your memory (ram/vram) can afford it (due to the quadratic complexity of attention), you likely want to enable this in your deployment too.
  - Almost concurrently, meta/facebook also released a paper that proposed basically the same technique. To their credit, they do acknowledge that the open source community independently come up with the same idea as they do.
  - Since then, some variants and extension of this idea have been proposed and even implemented: 1) **Dynamic interpolation** doesn't just scale 4x immediately, instead only scale as much as is needed based on the *current* context length as the text is generated, 2) **NTK aware scaled rope** uses a change of basis/nonlinear interpolation instead (NTK means "Neural Tangent Kernel"). 3) Dynamic NTK scaled rope. It is expected that the combination of these techniques, when combined with models fine-tuned for it (named SuperHOT due to quirks in history), would result in better output quality
  - LMSys released `LongChat` and `LongEval`, which tries to calm the scene by saying that although the model doesn't completely decohere, model capability do still seems to degrade towards but before the new limit at 8k is reached
- People are discussing the rumoured architecture of GPT4 (Mixture of Expert/MoE with 8x 220B), and someone proposed trying to replicate it using llama. One aggressive idea is that instead of training 8 separate models, one may try to train 8 separate fine-tunes.
- Another exciting news is a paper on **Classifier Free Guidance**, which is inspired from the corresponding technique in stable diffusion. If it works, it may allows one to "boost" the intelligence/capability of a LLM model up to 2x its size. So eg a 13B will behave more like a 30B after applying this. (At some additional compute cost but without increase in memory cost, and no retraining needed it seems)
- OpenAI said in a blog post that they believe ASI (Artificial Superintelligence) may arrive this decade. Keeping in mind that some are skeptical in that OpenAI has a vested interest and so may not be completely honest in what they truly believe, this might be a significant signal as this is a relatively aggressive estimates even after prediction market's estimate of AGI and ASI moving earlier in light of recent advances in AI in general. (Recall that ASI is more powerful than AGI and so should happen after it, and that their time will be close if you believe in a hard-takeoff/rapidly recursively self-improving scenario)

**Applications**

The two area that I have been personally looking at both seems to finally see some love this week:

- Someone made an open source "app-maker" using open LLM (WizardCoder)
- Another one carefully did the trial and error work to try to make open LLM work in an autonomous agent settings - `Nuggt`

If both of these actually work, I'd be excited too as these may be the first serious foray of open LLM into the LLM-based-app and more automation heavy part of the territory, which is a powerful but more tricky use case.
